include/envoy/request_info/request_info.h

```diff
@@ -71,63 +71,123 @@ class RequestInfo {
 
   /**
-   * @return the duration from request start to when the first byte of the response was received
-   * from the upstream host in microseconds.
+   * Set the request's protocol.
    */
-  virtual const Optional<std::chrono::microseconds>& responseReceivedDuration() const PURE;
+  virtual void protocol(Http::Protocol protocol) PURE;
 
   /**
-   * Set the duration from request start to when the first byte of the response was received
-   * from the upstream host in microseconds.
-   * @param time monotonic clock time when the first byte of the response was received.
+   * @return the response code.
    */
-  virtual void responseReceivedDuration(MonotonicTime time) PURE;
+  virtual const Optional<uint32_t>& responseCode() const PURE;
 
   /**
-   * @return the # of body bytes received in the request.
+   * @return the monotonic clock time when the last byte of the request was received.
    */
-  virtual uint64_t bytesReceived() const PURE;
+  virtual const Optional<MonotonicTime>& lastDownstreamRxByteReceived() const PURE;
 
   /**
-   * @return the protocol of the request.
+   * Set the monotonic clock time when the last byte of the request was received.
    */
-  virtual const Optional<Http::Protocol>& protocol() const PURE;
+  virtual void lastDownstreamRxByteReceived(MonotonicTime) PURE;
 
   /**
-   * Set the request's protocol.
+   * @return the monotonic clock time when the first byte of the request was sent upstream. There
+   * may be a considerable delta between lastDownstreamByteReceived and this value due to filters.
    */
-  virtual void protocol(Http::Protocol protocol) PURE;
+  virtual const Optional<MonotonicTime>& firstUpstreamTxByteSent() const PURE;
 
   /**
-   * @return the response code.
+   * Set the monotonic clock time when the first byte of the request was sent upstream.
    */
-  virtual const Optional<uint32_t>& responseCode() const PURE;
+  virtual void firstUpstreamTxByteSent(MonotonicTime) PURE;
 
   /**
-   * @return the # of body bytes sent in the response.
+   * @return the monotonic clock time when the last bye of the request was sent upstream.
    */
-  virtual uint64_t bytesSent() const PURE;
+  virtual const Optional<MonotonicTime>& lastUpstreamTxByteSent() const PURE;
+
+  /**
+   * Set the monotonic clock time when the last bye of the request was sent upstream.
+   */
+  virtual void lastUpstreamTxByteSent(MonotonicTime) PURE;
 
   /**
-   * @return the microseconds duration of the first byte received to the last byte sent.
+   * @return the monotonic clock time when the first byte of the response is received from upstream.
    */
-  virtual std::chrono::microseconds duration() const PURE;
+  virtual const Optional<MonotonicTime>& firstUpstreamRxByteReceived() const PURE;
+
+  /**
+   * Set the monotonic clock time when the first byte of the response is received from upstream.
+   */
+  virtual void firstUpstreamRxByteReceived(MonotonicTime) PURE;
```

Comment:
For all the upstream cases, we should clarify what we do in case of retries, where we've sent headers to the backend, it does an early failure, and we ship things off a second time.  I assume second pass gets recorded, but it's worth doccing and testing.